File: data-structures/javascript/graphs/directedAdjacencyMatrix.js
/*
 * Directed Adjacency Matrices are only good
 * when our graph is dense, meaning if
 * the number of edges is close to
 * the number of vertexes^2, or
 * the number of vertexes^2 is too less
 * to matter
 *
 * In the real world, i.e (Social Network,
 * Web Crawlers, etc...), Directed Adjacency
 * Matrices are inefficient. We are
 * better off using an Edge List,
 * or in other words, objects and
 * pointers, to represent our Graph
 *
 * Drawback could be that
 * we use an inefficient amount of
 * space with the Directed Adjacency Matrix
 * All the non edge indeces where
 * we have either Weight Infinity,
 * or a zero to signify it's not
 * a connected edge at the current
 * index, will take up more space
 * relative to the amount of actual
 * weighted/1-valued indeces in
 * our Directed Ajacency Matrix
 *
 * One difference between a DAG (Directed Acyclic Graph)
 * and a UDAG(Undirected Acyclic Graph)
 * is that the direction
 * of the edges is noted by 1 or -1.
 *
 * 1 = from This vertex draws/connects
 * and edge to another vertex
 *
 * -1 = another node
 * is drawing/connecting an edge
 * to the vertex with -1.
 */

class Stack {
  constructor() {
    this.items = [];
    this.size = 0;
  }

  insert(item) {
    this.items.push(item);
    this.size++;
  }

  pop() {
    this.size--;
    return this.items.pop();
  }

  contains(item) {
    return this.items.includes(item);
  }

  peek() {
    this.items[this.size - 1];
  }

  getMin() {
    return Math.min(...this.items);
  }

  getMax() {
    return Math.max(...this.items)
  }

  isEmpty() {
    return this.size === 0;
  }
}

class Queue {
  constructor() {
    this.items = [];
    this.size = 0;
  }

  enqueue(item) {
    this.items.push(item);
    this.size++;
  }

  dequeue(item) {
    this.size--;
    return this.items.shift();
  }

  contains(node) {
    return this.items.includes(node);
  }

  front() {
    return this.items[0];
  }

  back() {
    return this.items[this.items.length - 1];
  }

  printQueue() {
    return this.items;
  }

  queueSize() {
    return this.size;
  }

  max() {
    return Math.max(...this.items);
  }

  min() {
    return Math.min(...this.items);
  }

  isEmpty() {
    return this.size === 0;
  }
}

class Graph {
  constructor() {
    this.vertexList = {};
    this.vertexIndexList = {};
    this.nodesArray = [];
    this.adjacencyMatrix = [];
  }

  addVertex(vertex) {
    this.vertexList[this.nodesArray.length] = vertex;
    this.vertexIndexList[vertex] = this.nodesArray.length;
    this.nodesArray.push(vertex);

    for (let i = 0; i < this.adjacencyMatrix.length; i++) {
      this.adjacencyMatrix[i] =
        this.nodesArray
          .map((element, j) => {
            return !this.adjacencyMatrix[i][j] ?
              element = 0 :
              element = 1
            });
    }

    this.adjacencyMatrix.push(this.nodesArray.map(el => el = 0));
  }


  addEdge(fromVertex, toVertex) {
    this.adjacencyMatrix[this.vertexIndexList[fromVertex]][this.vertexIndexList[toVertex]] = 1;
    this.adjacencyMatrix[this.vertexIndexList[toVertex]][this.vertexIndexList[fromVertex]] = -1;
  }

  addEdges(fromVertex, toVertex) {
    this.adjacencyMatrix[this.vertexIndexList[fromVertex]][this.vertexIndexList[toVertex]] = 1;
    this.adjacencyMatrix[this.vertexIndexList[toVertex]][this.vertexIndexList[fromVertex]] = 1;
  }

  removeVertex(vertex) {
    let count = 0;
    let targetIndex = this.vertexIndexList[vertex];

    delete this.vertexIndexList[vertex];
    delete this.vertexList[targetIndex];

    this.nodesArray.splice(targetIndex, 1);

    this.adjacencyMatrix =
      this.adjacencyMatrix
        .filter((vertexEdges, i) => {
          return i !== targetIndex
        });

    for (let vertexEdges = 0; vertexEdges < this.adjacencyMatrix.length; vertexEdges++) {
      this.adjacencyMatrix[vertexEdges].splice(targetIndex, 1)
    }

    for (let vertex in this.vertexList) {
      if (count < this.nodesArray.length) {
        let temp = this.vertexList[vertex];
        this.vertexList[Number(vertex) - 1] = temp;
        this.vertexIndexList[this.vertexList[Number(vertex) - 1]]--;
      }
      count++;
    }

    delete this.vertexList[count];
  }

  removeEdge(fromVertex, toVertex) {
    this.adjacencyMatrix[this.vertexIndexList[fromVertex]][this.vertexIndexList[toVertex]] = 0;
    this.adjacencyMatrix[this.vertexIndexList[toVertex]][this.vertexIndexList[fromVertex]] = 0;
  }

  hasEdge(fromVertex, toVertex) {
    if (this.adjacencyMatrix[this.vertexIndexList[fromVertex]][this.vertexIndexList[toVertex]]) {
      return true;
    }

    return false;
  }

  DFSUtil(source_node, visited, stack){
    if (stack.size === this.nodesArray.length) {
      return;
    }

    if (!visited[source_node]) {
      stack.insert(source_node);
      visited[source_node] = true;
    }

    this.adjacencyMatrix[this.vertexIndexList[source_node]]
    .forEach((node, index) => {
      if(!visited[this.vertexList[index]] && node === 1){
        this.DFSUtil(this.vertexList[index], visited, stack);
      }
    });
  }

  depthFirstSearch(source_node) {
    let visited = {},
        stack   = new Stack();

    this.DFSUtil(source_node, visited, stack);
    return stack.items;
  }

  BFSUtil(source_node, visited, node_queue, result) {
    if (result.size === this.nodesArray.length) {
      return;
    }

    if(!visited[source_node]) {
      result.enqueue(source_node);
      visited[source_node] = true;
    }

    if (node_queue.contains(source_node)) {
      node_queue.dequeue();
    }

    this.adjacencyMatrix[this.vertexIndexList[source_node]]
    .forEach((node, index) => {
      if (!visited[this.vertexList[index]] && node === 1) {
        visited[this.vertexList[index]] = true;
        node_queue.enqueue(this.vertexList[index]);
        result.enqueue(this.vertexList[index]);
      }
    });

    node_queue.items.forEach((node, node_index) => {
      this.BFSUtil(node, visited, node_queue, result);
    });
  }

  breadthFirstSearch(source_node) {
    let visited = {},
        queue   = new Queue(),
        result  = new Queue();

    this.BFSUtil(source_node, visited, queue, result);
    return result.items;
  }

  hasCycleUtil(source_node, visited, visited_stack) {
    if (!visited[source_node]) {
      visited_stack.insert(source_node);
      visited[source_node] = true;

      let currentEdges = this.adjacencyMatrix[this.vertexIndexList[source_node]];

      for (let node = 0; node < currentEdges.length; node++) {
        if (currentEdges[node] === 1) {
          if (!visited[this.nodesArray[node]] && this.hasCycleUtil(this.nodesArray[node], visited, visited_stack)) {
            return true;
          } else if (visited_stack.contains(this.nodesArray[node])) {
            return true;
          }
        }
      }
    }

    visited_stack.pop();
    return false;
  }

  hasCycle() {
    let source_node = this.nodesArray[0];
    let visited = {};
    let visited_stack = new Stack();

    return this.hasCycleUtil(source_node, visited, visited_stack);
  }

  findCycleUtil(source_node, visited, visited_stack) {
    if (!visited[source_node]) {
      visited_stack.insert(source_node);
      visited[source_node] = true;

      let currentEdges = this.adjacencyMatrix[this.vertexIndexList[source_node]];

      for (let node = 0; node < currentEdges.length; node++) {
        if (currentEdges[node] === 1) {
          if (!visited[this.nodesArray[node]] && this.findCycleUtil(this.nodesArray[node], visited, visited_stack)) {
            return this.nodesArray[node];
          } else if (visited_stack.contains(this.nodesArray[node])) {
            return this.nodesArray[node];
          }
        }
      }
    }

    visited_stack.pop();
    return null;
  }

  findCycle() {
    let source_node = this.nodesArray[0];
    let visited = {};
    let visited_stack = new Stack();

    return this.findCycleUtil(source_node, visited, visited_stack)
  }

  topologicalSort() {
    if (this.hasCycle()) {
      throw new Error("Can't Topologically Sort a Cyclic Graph")
    }
    let vertexStack = new Stack();
    let nonDependentVertexes = [];
    let visited = {};

    this.nodesArray.forEach((node, index) => {
      if (this.adjacencyMatrix[this.vertexIndexList[node]].filter(edge => edge === 1).length >= 1) {
        visited[node] = true
        nonDependentVertexes.push(node);
      }
    });


    let current = nonDependentVertexes[nonDependentVertexes.length - 1];

    while(nonDependentVertexes.length) {
      let currentEdges = this.adjacencyMatrix[this.vertexIndexList[current]];
      currentEdges.forEach((edge, index) => {
        if (edge === 1 && !visited[this.vertexList[index]]) {
          vertexStack.insert(this.vertexList[index]);
          visited[this.vertexList[index]] = true;
        }
      });

      let removedNonDependentVertex = nonDependentVertexes.pop();
      vertexStack.insert(removedNonDependentVertex);
      visited[removedNonDependentVertex] = true;
      current = nonDependentVertexes[nonDependentVertexes.length - 1];
    }


    // at this point,
    // the stack has the
    // elements in proper order
    // however, for visual
    // purposes I will
    // push them into an array
    // essentially reversing it

    let result = [];
    let count = vertexStack.items.length;

    for (let i = 0; i < count; i++) {
      result.push(vertexStack.pop());
    }

    return result;
  }
}

let DAG = new Graph();

DAG.addVertex(10);
DAG.addVertex(12);
DAG.addVertex(14);
DAG.addVertex(16);
DAG.addVertex(100);

DAG.addEdge(10, 100);
DAG.addEdge(10, 12);
DAG.addEdge(12, 14);
console.log("\n\nNEW ACYCLIC GRAPH IS: ", DAG);
console.log("\n\n\nIF `false` then GRAPH IS ACYCLIC ELSE IT IS CYCLIC: ", DAG.hasCycle(), DAG.findCycle());


console.log("\n\nTOPOLOGICAL SORT: ", DAG.topologicalSort())

File: data-structures/go/graphs/directed_adjacency_matrix.go
package graphs

import (
  "errors"
  "google-interview-prep/data-structures/go/stacks"
  "google-interview-prep/data-structures/go/queues"
)

type DAM struct {
	Vertexes   [][]int
	VertexList []string
}

func ConstructDAM() *DAM {
	return &DAM{
		Vertexes:   make([][]int, 0),
		VertexList: make([]string, 0),
	}
}

func (g *DAM) AddVertex(node string) {
	g.Vertexes = append(g.Vertexes, make([]int, 0))
	for i, _ := range g.Vertexes {
		for len(g.Vertexes[i]) < len(g.Vertexes) {
			g.Vertexes[i] = append(g.Vertexes[i], 0)
		}
	}
	g.VertexList = append(g.VertexList, node)
}

func (g *DAM) Contains(node string) bool {
	for _, vert := range g.VertexList {
		if vert == node {
			return true
		}
	}

	return false
}

func (g *DAM) GetEdgeIndex(node string) int {
	fromIdx := 0
	for i, ver := range g.VertexList {
		if ver == node {
			fromIdx = i
		}
	}

	return fromIdx
}

func (g *DAM) RemoveVertex(node string) (bool, error) {
	v := false

	if g.Contains(node) {
		v = true
	} else {
		return false, errors.New("Vertex doesn't exist")
	}

	fromIdx := g.GetEdgeIndex(node)

	for _, vert := range g.Vertexes {
		for j, _ := range vert {
			g.RemoveEdges(g.VertexList[fromIdx], g.VertexList[j])
		}
	}

  for i, _ := range g.Vertexes {
    if fromIdx == 0 {
      g.Vertexes[i] = g.Vertexes[i][1:len(g.Vertexes[i])]
    } else if fromIdx == len(g.VertexList) - 1 {
      g.Vertexes[i] = g.Vertexes[i][0:len(g.Vertexes[i]) - 1]
    } else {
      g.Vertexes[i] = append(g.Vertexes[i][:fromIdx], g.Vertexes[i][fromIdx+1:]...)
    }
  }

  if fromIdx == 0 {
    g.VertexList = g.VertexList[1:len(g.VertexList)]
  } else if fromIdx == len(g.VertexList) - 1 {
    g.VertexList = g.VertexList[0:len(g.VertexList) - 1]
  } else {
    g.VertexList = append(g.VertexList[:fromIdx], g.VertexList[fromIdx+1:]...)
  }

	return v, nil
}

func (g *DAM) HasEdge(fromNode, toNode string) bool {
	fromIdx, toIdx := g.GetEdgeIndex(fromNode), g.GetEdgeIndex(toNode)

	if g.Vertexes[fromIdx][toIdx] == 1 {
		return true
	}

	return false
}

func (g *DAM) SameNodes(fromNode, toNode string) bool {
	return fromNode == toNode
}

func (g *DAM) AddEdge(fromNode, toNode string) {
	fromIdx, toIdx := g.GetEdgeIndex(fromNode), g.GetEdgeIndex(toNode)
	g.Vertexes[fromIdx][toIdx] = 1
}

func (g *DAM) AddEdges(fromNode, toNode string) {
  fromIdx, toIdx := g.GetEdgeIndex(fromNode), g.GetEdgeIndex(toNode)
  g.Vertexes[fromIdx][toIdx] = 1
  g.Vertexes[toIdx][fromIdx] = -1
}

func (g *DAM) RemoveEdge(fromNode, toNode string) {
	fromIdx, toIdx := g.GetEdgeIndex(fromNode), g.GetEdgeIndex(toNode)
	g.Vertexes[fromIdx][toIdx] = 0
}

func (g *DAM) RemoveEdges(fromNode, toNode string) {
	g.RemoveEdge(fromNode, toNode)
	g.RemoveEdge(toNode, fromNode)
}

func (g *DAM) hasCycleUtil(sourceNode string, visited map[string]bool, visitedStack *stacks.StringStack) bool {
  if !visited[sourceNode] {
    visitedStack.Push(sourceNode)
    visited[sourceNode] = true
    fromIdx := g.GetEdgeIndex(sourceNode)
    currentEdges := g.Vertexes[fromIdx]

    for node := 0; node < len(currentEdges); node++ {
      if currentEdges[node] == 1 {
        if visited[g.VertexList[node]] == false && g.hasCycleUtil(g.VertexList[node], visited, visitedStack) == true {
          return true
        } else if visitedStack.Contains(g.VertexList[node]) == true {
          return true
        }
      }
    }
  }

  visitedStack.Pop()
  return false
}

func (g *DAM) HasCycle() bool {
  sourceNode := g.VertexList[0]
  visited := make(map[string]bool)
  visitedStack := &stacks.StringStack{
      Items: make([]string, 0),
      Size: 0,
  }

  return g.hasCycleUtil(sourceNode, visited, visitedStack);
}

func (g *DAM) DFS(fromNode string) []string {
	s := &stacks.StringStack{
		Items: make([]string, 0),
		Size:  0,
	}
	seen := make(map[string]bool)

	s = g.DepthFistSearch(fromNode, s, seen)
	return s.Items
}

func (g *DAM) DepthFistSearch(node string, s *stacks.StringStack, seen map[string]bool) *stacks.StringStack {
	if s.Size == len(g.Vertexes) {
		return s
	}

	if !seen[node] {
		seen[node] = true
		s.Push(node)
	}

	fromIdx := g.GetEdgeIndex(node)

	for i, _ := range g.Vertexes[fromIdx] {
		if seen[g.VertexList[i]] == false && g.Vertexes[fromIdx][i] != 0 {
			s = g.DepthFistSearch(g.VertexList[i], s, seen)
		}
	}

	return s
}

func (g *DAM) BFS(fromNode string) []string {
	q := &queues.StringQueue{
		Items: make([]string, 0),
		Size:  0,
	}
	s := &stacks.StringStack{
		Items: make([]string, 0),
		Size:  0,
	}
	seen := make(map[string]bool)

	s = g.BreadthFirstSearch(fromNode, q, s, seen)
	return s.Items
}

func (g *DAM) BreadthFirstSearch( node string, q *queues.StringQueue, s *stacks.StringStack, seen map[string]bool) *stacks.StringStack {
	if s.Size == len(g.Vertexes) {
		return s
	}

	if !seen[node] {
		seen[node] = true
		s.Push(node)
	}

	fromIdx := g.GetEdgeIndex(node)

	for i, _ := range g.Vertexes[fromIdx] {
		if !seen[g.VertexList[i]] && g.Vertexes[fromIdx][i] != 0 {
			q.Enqueue(g.VertexList[i])
		}
	}

	for q.Size != 0 {
		popped := q.Dequeue()
		s = g.BreadthFirstSearch(popped, q, s, seen)
	}

	return s
}
